source/agora/test/Quorum.d

```diff
@@ -0,0 +1,157 @@
+import std.format;
+import std.range;
+
+import core.thread;
+import core.time;
+
+///
+unittest
+{
+    // generate 1006 blocks, 2 short of the enrollments expiring.
+    TestConf conf = { nodes : 4, max_listeners : 5,
+        topology : NetworkTopology.TwoOutsiderValidators, extra_blocks : 1006 };
+    auto network = makeTestNetwork(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto nodes = network.clients;
+
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getBlockHeight() == 1006, 5.seconds,
+            format("Node %s has block height %s. Expected: %s",
+                idx, node.getBlockHeight(), 1006)));
+
+    Transaction makePayTx (in Transaction prev, PublicKey[] addresses, uint index = 0)
+    {
+        auto input = Input(hashFull(prev), index);
+
+        Transaction tx =
+        {
+            TxType.Payment,
+            [input],
+            addresses.map!(addr => Output(Amount.MinFreezeAmount, addr)).array
+        };
+
+        auto signature = getGenesisKeyPair().secret.sign(hashFull(tx)[]);
+        tx.inputs[0].signature = signature;
+        return tx;
+    }
+
+    Transaction makeFreezeTransaction (in Transaction prev, PublicKey address)
+    {
+        auto input = Input(hashFull(prev), 0);
+
+        Transaction tx =
+        {
+            TxType.Freeze,
+            [input],
+            [Output(Amount.MinFreezeAmount, address)]
+        };
+
+        auto signature = getGenesisKeyPair().secret.sign(hashFull(tx)[]);
+        tx.inputs[0].signature = signature;
+        return tx;
+    }
+
+    // create block with 6 payment and 2 freeze tx's
+    auto txs = makeChainedTransactions(getGenesisKeyPair(),
+        network.blocks[$ - 1].txs, 1);
+
+    // rewrite 3rd to last tx to multiple outputs so we can create 8 spend tx's
+    // in next block
+    txs[$ - 3] = makePayTx(network.blocks[$ - 1].txs[$ - 3],
+        [getGenesisKeyPair.address, getGenesisKeyPair.address,
+        getGenesisKeyPair.address]);
+
+    // rewrite the last two tx's to be freeze tx's for our outsider validator nodes
+    txs[$ - 2] = makeFreezeTransaction(network.blocks[$ - 1].txs[$ - 2],
+        nodes[$ - 2].getPublicKey());
+    txs[$ - 1] = makeFreezeTransaction(network.blocks[$ - 1].txs[$ - 1],
+        nodes[$ - 1].getPublicKey());
+    txs.each!(tx => nodes[0].putTransaction(tx));
+
+    // at block height 1007 the freeze tx's are available
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getBlockHeight() == 1007, 5.seconds,
+            format("Node %s has block height %s. Expected: %s",
+                idx, node.getBlockHeight(), 1007)));
+
+    // now we can create enrollments
+    Enrollment enroll_0 = nodes[$ - 2].createEnrollmentData();
+    Enrollment enroll_1 = nodes[$ - 1].createEnrollmentData();
+    nodes[2].enrollValidator(enroll_0);
+    nodes[3].enrollValidator(enroll_1);
```

Comment:
Question about this: Should we add the ability to enroll as soon as we freeze ?
Pro:
- Genesis does it;
- It allows to potentially solve the censorship issue if paired with a rule that more validators => Higher priority;
Cons:
- Attack vector ?

Reply:
Let's file an issue for this. And yes, I wanted this feature too.